### ISSUE TITLE
Full push scoping while Sidecar CRD was configured

### ISSUE THREAD

**Issue by seflerZ** [istio/istio#22392]
**Describe the feature request**
I'm totally confused that while Sidecar enabled push data scoping for proxy, the full push will still be triggered not matter what the service is. For example, If proxy A was configured with service-A\B\C, it will received only the data of those services. However, if CDS of service-D was changed, the full push for proxy A still got triggered and result in full push of service-A\B\C to it. Even though there is nothing changed among those CDS!

I think we can do simple scoping while doing full push. This will be simple as if we add some context information in the push event.

**Describe alternatives you've considered**
None

**Additional context**

**Comment by howardjohn** [comment]
We should already be doing this: https://github.com/istio/istio/blob/f2d9e32987c4dd10aac16738c3072eaadf6377b4/pilot/pkg/proxy/envoy/v2/ads_common.go#L22. It may need to be expanded to cover more cases, but it is intended to handle what you have described

**Comment by seflerZ** [comment]
Yes, my friend. I've checked that code and found only namespace applied. I believe we can do some simple modification to extend this to services level.

**Comment by howardjohn** [comment]
Ah I see, we do it at namespace level. Thanks

**Comment by seflerZ** [comment]
The idea would be:

Add extra configure resource names changed to "pushEv *XdsEvent"

```
type XdsEvent struct {
	// If not empty, it is used to indicate the event is caused by a change in the clusters.
	// Only EDS for the listed clusters will be sent.
	- edsUpdatedServices map[string]struct{}

	namespacesUpdated map[string]struct{}

	-  configTypesUpdated map[resource.GroupVersionKind]struct{}
        + configsUpdated map[resource.GroupVersionKind]map[string]interface{}
        ...
}
```

We can use the "configsUpdated" attribute to hold all the changes, including "edsUpdatedServices", "configTypesUpdated" and stores interface like:
```
type ChangedResource interface {
    Merge(o1, o2)
}
``` 
In the code, we can cast the resource to the one the "GroupVersionKind" represents and access any thing you want after that.

**Comment by YonkaFang** [comment]
Basing on the merged PR #22495 , we'd like to continue dealing with the full-push scoping issue.

As we already have all updated configs/resources in `PushRequest.ConfigsUpdated`  field, we can have a component distinguishing whether changes of resources of a `GroupVersionKind` will affect a specified proxy:

```go
var (
    pushResourceScope = map[GroupVersionKind]func(proxy *model.Proxy, pushEv *XdsEvent, resources map[string]struct{}) bool {}
    defaultTrueScope = func(proxy *model.Proxy, pushEv *XdsEvent, resources map[string]struct{}) bool {
        return true
    }
    defaultFalseScope = func(proxy *model.Proxy, pushEv *XdsEvent, resources map[string]struct{}) bool {
        return false
    }
)

// Initialize it explicitly or leave it to be filling in relevant-business module's init stage.

func PushAffectProxy(pushEv *XdsEvent, proxy *model.Proxy) bool {
  if len(pushEv.configsUpdated) == 0 {
    return true
  }

  for kind, resources := range pushEv.configsUpdated {
    if scope, f := pushResourceScope[kind]; !f {
      return true
    } else if scope(proxy, pushEv, resources) {
      return true
    }
  }

  return false
}

func init() {
  // Take ServiceEntry as example:
  pushResourceScope[ServiceEntryKind] = func(proxy *model.Proxy, pushEv *XdsEvent, resources map[string]struct{}) bool {
    for _, resource := range resources {
      if proxy.SidecarScope.DependsOnService(resource) {
        return true
      }
    }

    return false
  }

  // Assume that we can not tell whether a Sidecar resource will affect a proxy, or leave to impl. 
  pushResourceScope[SidecarKind] = defaultTrueScope

  // etc ...
}

func (sc *SidecarScope) DependsOnService(service string) bool {
	if sc == nil {
		return true
	}

	if _, f := sc.NamespaceForHostname[host.Name(service)]; f {
		return true
	}

	return false
}
```

This scoping mechanism can be used in `ProxyNeedsPush` and we'll get a result that: A full-push will be performed towards a proxy only when it cares about the updated configs.

@howardjohn Have any idea?

**Comment by seflerZ** [comment]
@rshriram How do you think about this? This will avoid lots of unnecessary pushes.
